apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,476 @@
+        inputNode.addEventListener('focus', this.onFocus);
+        inputNode.addEventListener('blur', this.onBlur);
+        inputNode.addEventListener('input', this.onInput);
+      });
+
+      // Find (when cached HTML) or create list to hold the autocomplete matches
+      // Put the list inside the autocomplete element instead of off the
+      // document, as it should be within a given UI card, in case in-app
+      // overlay UIs like confirm dialogs need to be shown.
+      this.list = this.querySelector('.cmp-autocomplete-list');
+      if (!this.list) {
+        this.list = document.createElement('ul');
+        this.list.classList.add('cmp-autocomplete-list');
+        this.appendChild(this.list);
+      }
+      this.list.addEventListener('click', this.onMatchClick.bind(this));
+      this.list.innerHTML = '';
+      this.list.classList.add('collapsed');
+
+      // A spacer is used to allow scrolling of the parentNode so the target
+      // input is at the top of the view.
+      this.spacer = this.querySelector('.cmp-autocomplete-spacer');
+      if (!this.spacer) {
+        this.spacer = document.createElement('div');
+        this.spacer.classList.add('cmp-autocomplete-spacer');
+        this.appendChild(this.spacer);
+      }
+
+      // Hold on to the element that holds all the input fields, to use it for
+      // sizing and setting scroll behavior amongst the inputs while the
+      // autocomplete is active.
+      this.inputArea = this.parentNode
+                       .querySelector('.cmp-autocomplete-input-list');
+
+      // Find parent that is scrollable and origin for measurements.
+      var parentNode = this;
+      this.originTop = 0;
+      while((parentNode = parentNode.parentNode)) {
+        if (parentNode.classList.contains('cmp-autocomplete-origin')) {
+          this.originNode = parentNode;
+          break;
+        }
+      }
+
+      // Set up data source.
+      var dataSourceId = this.dataset.source;
+      if (dataSourceId) {
+        this.bindDataSource(dataSourceId);
+      }
+    },
+
+    /**
+     * Custom element callback when element is placed in the document.
+     */
+    attachedCallback: function() {
+      this.onResize = this.onResize.bind(this);
+      window.addEventListener('resize', this.onResize, false);
+    },
+
+    /**
+     * Custom element callback when element is removed from the document.
+     */
+    detachedCallback: function() {
+      window.removeEventListener('resize', this.onResize, false);
+    },
+
+    /**
+     * Triggered by input focus events.
+     */
+    onFocus: function(event) {
+      this.inputNode = event.target;
+      this.query = this.inputNode.value.trim();
+      this.match();
+    },
+
+    /**
+     * Triggered by input blur events.
+     */
+    onBlur: function(event) {
+      this._blurRequest = Date.now();
+      // Taps on the autocomplete list can trigger a blur, so give those taps
+      // a chance to emit events before hiding the list.
```

Comment:
Oy.  I guess I'm not surprised that it's necessary then.  I was hoping you were just being superstitious (I assumed this because I've literally written logic exactly like this just because I didn't want to think about/debug the problem.)

A zero setTimeout would be fine-ish with me with a little more comment along the lines of "the blur event is dispatched before the click event.  These are dispatched consecutively, so we want to run immediately after any click event would have been dispatched. We can't use a Promise because the micro-task queue will be drained before the click event is processed."